File: 101-The Blocks Problem/band.cpp
#include <iostream>
#include <list>
#include <string>
#include <boost/algorithm/string.hpp>
#include <vector>
#include <algorithm>
using namespace std;
using namespace boost;


void printBlocks(vector<list<int>>, int size);
void band(vector<list<int>>*, int);
int findBlocks(vector<list<int>>, int, int);
void moveOnto(vector<list<int>>*, int, int, int);
void moveOver(vector<list<int>>*, int, int, int);
void pileOnto(vector<list<int>>*, int, int, int);
void pileOver(vector<list<int>>*, int, int, int);


int main()
{
    int size;    cin>>size;     cin.ignore();

    vector<list<int>> blocks;

    for(int a=0; a<size; a++)
    {
        list<int> temp;     temp.push_back(a);      blocks.push_back(temp);
    }

    band(&blocks, size);

    printBlocks(blocks, size);

    return 0;
}


void printBlocks(vector<list<int>> blocks, int size)
{
    for(int a=0; a<size; a++)
    {
        list<int> temp;     temp = blocks[a];      cout<<a<<":";
        
        while(!temp.empty())
        {
            cout<<" "<<temp.front();    temp.pop_front();
        }

        cout<<endl;
    }

    return;
}


void band(vector<list<int>>* blocks, int size)
{
    string input;
    while (1)
    {
        getline(cin, input);

        if(input == "quit")
        {
            return;
        }
        else
        {
            vector<string> output;
            split(output, input, is_any_of(" "));

            if(output[1] != output[3] and findBlocks(blocks[0], size, stoi(output[1])) != findBlocks(blocks[0], size, stoi(output[3])))
            {
                if(output[0] == "move")
                {
                    if (output[2] == "onto")
                    {
                        moveOnto(blocks, size, stoi(output[1]), stoi(output[3]));
                    }
                    else if(output[2] == "over")
                    {
                        moveOver(blocks, size, stoi(output[1]), stoi(output[3]));
                    }
                }
                else if(output[0] == "pile")
                {
                    if (output[2] == "onto")
                    {
                        pileOnto(blocks, size, stoi(output[1]), stoi(output[3]));
                    }
                    else if(output[2] == "over")
                    {
                        pileOver(blocks, size, stoi(output[1]), stoi(output[3]));
                    }
                }
            }
        }
    }

    return;
}


void moveOnto(vector<list<int>>* blocks, int size, int a, int b)
{    
    if(!blocks[0][a].empty())
    {
        while(!blocks[0][a].empty())
        {
            if(blocks[0][a].front() != a)
            {
                blocks[0][blocks[0][a].front()].push_back(blocks[0][a].front());    blocks[0][a].pop_front();
            }
            else
            {
                blocks[0][a].pop_front();
            }
        }
    }
    else
    {
        int index = findBlocks(blocks[0], size, a);
        
        list<int> temp;
        while(blocks[0][index].front() != a)
        {
            temp.push_back(blocks[0][index].front());   blocks[0][index].pop_front(); 
        }
        blocks[0][index].pop_front();

        while(!blocks[0][index].empty())
        {
            blocks[0][blocks[0][index].front()].push_back(blocks[0][index].front());    blocks[0][index].pop_front(); 
        }

        blocks[0][index] = temp;
    }
    
    if(!blocks[0][b].empty())
    {
        while(!blocks[0][b].empty())
        {
            if(blocks[0][b].front() != b)
            {
                blocks[0][blocks[0][b].front()].push_back(blocks[0][b].front());    blocks[0][b].pop_front();
            }
            else
            {
                blocks[0][b].pop_front();
            }
        }

        blocks[0][b].push_back(b);      blocks[0][b].push_back(a);
    }
    else
    {
        int index = findBlocks(blocks[0], size, b);

        list<int> temp;
        while(blocks[0][index].front() != b)
        {
            temp.push_back(blocks[0][index].front());   blocks[0][index].pop_front(); 
        }
        temp.push_back(blocks[0][index].front());   blocks[0][index].pop_front();

        while(!blocks[0][index].empty())
        {
            blocks[0][blocks[0][index].front()].push_back(blocks[0][index].front());    blocks[0][index].pop_front(); 
        }

        blocks[0][index] = temp;    blocks[0][index].push_back(a);
    }

    return;
}


int findBlocks(vector<list<int>> blocks, int size, int b)
{
    for(int a=0; a<size ;a++)
    {
        list<int>::iterator it = find(blocks[a].begin(), blocks[a].end(), b);
        if(blocks[a].end() != it)
        {
            return a;
        }        
    }

    return -1;
}


void moveOver(vector<list<int>>* blocks, int size, int a, int b)
{
    if(!blocks[0][a].empty())
    {
        while(!blocks[0][a].empty())
        {
            if(blocks[0][a].front() != a)
            {
                blocks[0][blocks[0][a].front()].push_back(blocks[0][a].front());    blocks[0][a].pop_front();
            }
            else
            {
                blocks[0][a].pop_front();
            }
        }
    }
    else
    {
        int index = findBlocks(blocks[0], size, a);
        
        list<int> temp;
        while(blocks[0][index].front() != a)
        {
            temp.push_back(blocks[0][index].front());   blocks[0][index].pop_front(); 
        }
        blocks[0][index].pop_front();

        while(!blocks[0][index].empty())
        {
            blocks[0][blocks[0][index].front()].push_back(blocks[0][index].front());    blocks[0][index].pop_front(); 
        }

        blocks[0][index] = temp;
    }
    
    if(!blocks[0][b].empty())
    {
        blocks[0][b].push_back(a);
    }
    else
    {
        blocks[0][findBlocks(blocks[0], size, b)].push_back(a);
    }

    return;
}


void pileOnto(vector<list<int>>* blocks, int size, int a, int b)
{
    if(!blocks[0][b].empty())
    {
        while(!blocks[0][b].empty())
        {
            if(blocks[0][b].front() != b)
            {
                blocks[0][blocks[0][b].front()].push_back(blocks[0][b].front());    blocks[0][b].pop_front();
            }
            else
            {
                blocks[0][b].pop_front();
            }
        }

        blocks[0][b].push_back(b);
    }
    else
    {
        int index = findBlocks(blocks[0], size, b);

        list<int> temp;
        while(blocks[0][index].front() != b)
        {
            temp.push_back(blocks[0][index].front());   blocks[0][index].pop_front();
        }
        temp.push_back(blocks[0][index].front());   blocks[0][index].pop_front();

        while(!blocks[0][index].empty())
        {
            blocks[0][blocks[0][index].front()].push_back(blocks[0][index].front());    blocks[0][index].pop_front(); 
        }

        blocks[0][index] = temp;    b = index;
    }

    if(!blocks[0][a].empty())
    {
        while(!blocks[0][a].empty())
        {
            blocks[0][b].push_back(blocks[0][a].front());  blocks[0][a].pop_front();
        }
    }
    else
    {
        int index = findBlocks(blocks[0], size, a);

        list<int> temp;
        while(blocks[0][index].front() != a)
        {
            temp.push_back(blocks[0][index].front());   blocks[0][index].pop_front(); 
        }

        while(!blocks[0][index].empty())
        {
            blocks[0][b].push_back(blocks[0][index].front());  blocks[0][index].pop_front();
        }
        
        blocks[0][index] = temp;
    }
    
    return;
}


void pileOver(vector<list<int>>* blocks, int size, int a, int b)
{ 
    if(blocks[0][b].empty())
    {
        b = findBlocks(blocks[0], size, b);
    }
    
    if(!blocks[0][a].empty())
    {
        while(!blocks[0][a].empty())
        {
            blocks[0][b].push_back(blocks[0][a].front());       blocks[0][a].pop_front();
        }
    }
    else
    {
        int index = findBlocks(blocks[0], size, a);

        list<int> temp;
        while(blocks[0][index].front() != a)
        {
            temp.push_back(blocks[0][index].front());   blocks[0][index].pop_front(); 
        }

        while(!blocks[0][index].empty())
        {
            blocks[0][b].push_back(blocks[0][index].front());       blocks[0][index].pop_front();
        }

        blocks[0][index] = temp;
    }

    return;
}

File: 101-The Blocks Problem/band.py
def printBlocks(blocks, size):
    for a in range(size):
        print(a, end='')
        print(':', end = '')
        temp = blocks[a]
        if temp != []:
            for b in temp:
                print('', b, end='')
        print(end='\n')

    return


def band(blocks):
    while 1:
        line = input()

        if line == 'quit':
            break
            return
        lint = line.split(' ', 4)
        if (int(lint[1]) != int(lint[3])) and (findBlock(blocks, int(lint[1])) != findBlock(blocks, int(lint[3]))):
            if lint[0] == 'move':
                if lint[2] == 'onto':
                    moveOnto(blocks, int(lint[1]), int(lint[3]))
                elif lint[2] == 'over':
                    moveOver(blocks, int(lint[1]), int(lint[3]))
            elif lint[0] == 'pile':
                if lint[2] == 'onto':
                    pileOnto(blocks, int(lint[1]), int(lint[3]))
                elif lint[2] == 'over':
                    pileOver(blocks, int(lint[1]), int(lint[3]))

    return


def moveOnto(blocks, a, b):
    if blocks[a] != []:
        for c in blocks[a]:
            if c != a:
                blocks[c].append(c)

        blocks[a] = []
    else:
        index = findBlock(blocks, a)
        temp = blocks[index]
        temp_1 = []
        while 1:
            if temp[0] != a:
                temp_1.append(temp.pop(0))
            else:
                break
        blocks[index] = temp_1
        temp.pop(0)
        
        for c in temp:
            blocks[c].append(c)
    
    if blocks[b] != []:
        for c in blocks[b]:
            if c != b:
                blocks[c].append(c)

        blocks[b] = []
        blocks[b].append(b)
        blocks[b].append(a)
    else:
        index = findBlock(blocks, b)
        temp = blocks[index]
        temp_1 = []
        while 1:
            if temp[0] != b:
                temp_1.append(temp.pop(0))
            else:
                break
        temp_1.append(temp.pop(0))

        for c in temp:
            blocks[c].append(c)

        temp_1.append(a)
        blocks[index] = temp_1

    return


def moveOver(blocks, a, b):
    if blocks[a] != []:
        for c in blocks[a]:
            if c != a:
                blocks[c].append(c)

        blocks[a] = []
    else:
        index = findBlock(blocks, a)
        temp = blocks[index]
        temp_1 = []
        while 1:
            if temp[0] != a:
                temp_1.append(temp.pop(0))
            else:
                break  
        temp.pop(0)
        blocks[index] = temp_1
        
        for c in temp:
            blocks[c].append(c)

    index = b
    if blocks[index] == []:
        index = findBlock(blocks, b)

    blocks[index].append(a)

    return


def pileOnto(blocks, a, b):
    if blocks[b] != []:
        for c in blocks[b]:
            if c != b:
                blocks[c].append(c)

        blocks[b] = []
        blocks[b].append(b)
    else:
        index = findBlock(blocks, b)
        temp = blocks[index]
        temp_1 = []
        while 1:
            if temp[0]!=b:
                temp_1.append(temp.pop(0))
            else:
                break
        temp_1.append(temp.pop(0))

        for c in temp:
            blocks[c].append(c)
        
        blocks[index] = temp_1
        b = index

    if blocks[a] != []:
        for c in blocks[a]:
            blocks[b].append(c)

        blocks[a] = []
    else:
        index = findBlock(blocks, a)
        temp = blocks[index]
        temp_1 = []
        while 1:
            if temp[0] != a:
                temp_1.append(temp.pop(0))
            else:
                break
        
        blocks[index] = temp_1

        for c in temp:
            blocks[b].append(c)

    return


def pileOver(blocks, a, b):
    if blocks[a] != []:
        temp = blocks[a]
        blocks[a] = []
    else:
        index = findBlock(blocks, a)
        temp = blocks[index]
        temp_1 = []
        while 1:
            if temp[0] != a:
                temp_1.append(temp.pop(0))
            else:
                break

        blocks[index] = temp_1

    if blocks[b] != []:
        for c in temp:
            blocks[b].append(c)
    else:
        index = findBlock(blocks, b)
        for c in temp:
            blocks[index].append(c)

    return


def findBlock(blocks, a):
    index  = 0

    for b in blocks:
        for c in b:
            if c == a:
                return index
        
        index += 1


if __name__ == "__main__":
    blocks = []
    size = int(input())

    for a in range(size):
        temp = []
        temp.append(a)
        blocks.append(temp)
    
    band(blocks)
    printBlocks(blocks, size)

    exit(0)
